pay/contracts/subscription/HectorSubscription.sol

```diff
@@ -0,0 +1,498 @@
+        uint256 refundForOldPlan,
+        uint256 indexed newPlanId,
+        uint256 payForNewPlan,
+        uint48 expiredAt
+    );
+    event PayerDeposit(
+        address indexed from,
+        address indexed token,
+        uint256 amount
+    );
+    event PayerWithdraw(
+        address indexed from,
+        address indexed token,
+        uint256 amount
+    );
+    event Refunded(address indexed to, address indexed token, uint256 amount);
+
+    /* ======== INITIALIZATION ======== */
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize() external initializer {
+        Factory factory = Factory(msg.sender);
+
+        (product, treasury) = abi.decode(
+            factory.parameter(),
+            (string, address)
+        );
+
+        plans.push(Plan({token: address(0), period: 0, amount: 0}));
+
+        moderators[factory.owner()] = true;
+
+        _transferOwnership(factory.owner());
+        __ReentrancyGuard_init();
+    }
+
+    /* ======== MODIFIER ======== */
+
+    modifier onlyMod() {
+        if (!moderators[msg.sender]) revert INVALID_MODERATOR();
+        _;
+    }
+
+    modifier onlyValidPlan(uint256 _planId) {
+        if (_planId == 0 || _planId >= plans.length) revert INVALID_PLAN();
+        _;
+    }
+
+    /* ======== INTERNAL FUNCTIONS ======== */
+
+    function max(uint256 a, uint256 b) internal pure returns (uint256) {
+        return a >= b ? a : b;
+    }
+
+    function min(uint256 a, uint256 b) internal pure returns (uint256) {
+        return a < b ? a : b;
+    }
+
+    /* ======== POLICY FUNCTIONS ======== */
+
+    function setTreasury(address _treasury) external onlyOwner {
+        if (_treasury == address(0)) revert INVALID_ADDRESS();
+        treasury = _treasury;
+    }
+
+    function appendPlan(Plan[] calldata _plans) external onlyOwner {
+        uint256 length = _plans.length;
+        for (uint256 i = 0; i < length; i++) {
+            Plan memory _plan = _plans[i];
+
+            if (_plan.token == address(0)) revert INVALID_ADDRESS();
+            if (_plan.period == 0) revert INVALID_TIME();
+            if (_plan.amount == 0) revert INVALID_AMOUNT();
+
+            plans.push(_plan);
+
+            emit PlanUpdated(
+                plans.length - 1,
+                _plan.token,
+                _plan.period,
+                _plan.amount
+            );
+        }
+    }
+
+    function updatePlan(uint256 _planId, Plan calldata _plan)
+        external
+        onlyOwner
+        onlyValidPlan(_planId)
+    {
+        if (_plan.token == address(0)) revert INVALID_ADDRESS();
+        if (_plan.period == 0) revert INVALID_TIME();
+        if (_plan.amount == 0) revert INVALID_AMOUNT();
+
+        plans[_planId] = _plan;
+
+        emit PlanUpdated(_planId, _plan.token, _plan.period, _plan.amount);
+    }
+
+    function setModerator(address _moderator, bool _approved)
```

Comment:
@bejaxer we also need a func to remove a list of moderators

Reply:
approved: true => add
approved: false => remove